compiler/noirc_evaluator/src/ssa/opt/inlining.rs

```diff
@@ -113,28 +123,121 @@ struct PerFunctionContext<'function> {
     inlining_entry: bool,
 }
 
-/// The entry point functions are each function we should inline into - and each function that
-/// should be left in the final program.
-/// This is the `main` function, any Acir functions with a [fold inline type][InlineType::Fold],
-/// and any brillig functions used.
-fn get_entry_point_functions(
+/// Utility function to find out the direct calls of a function.
+fn called_functions(func: &Function) -> BTreeSet<FunctionId> {
+    let mut called_function_ids = BTreeSet::default();
+    for block_id in func.reachable_blocks() {
+        for instruction_id in func.dfg[block_id].instructions() {
+            let Instruction::Call { func: called_value_id, .. } = &func.dfg[*instruction_id] else {
+                continue;
+            };
+
+            if let Value::Function(function_id) = func.dfg[*called_value_id] {
+                called_function_ids.insert(function_id);
+            }
+        }
+    }
+
+    called_function_ids
+}
+
+// Recursively explore the SSA to find the functions that end up calling themselves
+fn find_recursive_functions(
+    ssa: &Ssa,
+    current_function: FunctionId,
+    mut explored_functions: ImmutableHashSet<FunctionId>,
+    recursive_functions: &mut BTreeSet<FunctionId>,
+) {
+    if explored_functions.contains(&current_function) {
+        recursive_functions.insert(current_function);
+        return;
+    }
+
+    let called_functions = called_functions(
+        ssa.functions.get(&current_function).expect("Function should exist in SSA"),
+    );
+
+    explored_functions.insert(current_function);
+
+    for called_function in called_functions {
+        find_recursive_functions(
+            ssa,
+            called_function,
+            explored_functions.clone(),
+            recursive_functions,
+        );
+    }
+}
+
+fn find_all_recursive_functions(ssa: &Ssa) -> BTreeSet<FunctionId> {
+    let mut recursive_functions = BTreeSet::default();
+    find_recursive_functions(
+        ssa,
+        ssa.main_id,
+        ImmutableHashSet::default(),
+        &mut recursive_functions,
+    );
+    recursive_functions
+}
+
+/// The functions we should inline into (and that should be left in the final program) are:
+///  - main
+///  - Any Brillig function called from Acir
+///  - Any Brillig recursive function (Acir recursive functions will be inlined into the main function)
+///  - Any Acir functions with a [fold inline type][InlineType::Fold],
+fn get_functions_to_inline_into(
     ssa: &Ssa,
     no_predicates_is_entry_point: bool,
 ) -> BTreeSet<FunctionId> {
-    let functions = ssa.functions.iter();
-    let mut entry_points = functions
+    let brillig_entry_points: BTreeSet<_> = ssa
```

Comment:
Let's combine this with the acir_entry_points loop below to avoid looping twice. If it's easier we can change it to a for loop instead